iroh-bytes/src/baomap.rs

```diff
@@ -0,0 +1,288 @@
+
+    /// Upgrade a partial entry to a complete entry.
+    fn insert_complete(&self, entry: Self::PartialEntry) -> BoxFuture<'_, io::Result<()>>;
+}
+
+/// Extension of BaoMap to add misc methods used by the rpc calls.
+pub trait ReadonlyStore: Map {
+    /// list all blobs in the database. This should include collections, since
+    /// collections are blobs and can be requested as blobs.
+    fn blobs(&self) -> Box<dyn Iterator<Item = Hash> + Send + Sync + 'static>;
+    /// list all roots (collections or other explicitly added things) in the database
+    fn roots(&self) -> Box<dyn Iterator<Item = Hash> + Send + Sync + 'static>;
+    /// Validate the database
+    fn validate(&self, tx: mpsc::Sender<ValidateProgress>) -> BoxFuture<'_, anyhow::Result<()>>;
+
+    /// list partial blobs in the database
+    fn partial_blobs(&self) -> Box<dyn Iterator<Item = Hash> + Send + Sync + 'static>;
+
+    /// This trait method extracts a file to a local path.
+    ///
+    /// `hash` is the hash of the file
+    /// `target` is the path to the target file
+    /// `mode` is a hint how the file should be exported.
+    /// `progress` is a callback that is called with the total number of bytes that have been written
+    fn export(
+        &self,
+        hash: Hash,
+        target: PathBuf,
+        mode: ExportMode,
+        progress: impl Fn(u64) -> io::Result<()> + Send + Sync + 'static,
+    ) -> BoxFuture<'_, io::Result<()>>;
+}
+
+/// The mutable part of a BaoDb
+pub trait Store: ReadonlyStore + PartialMap {
+    /// This trait method imports a file from a local path.
+    ///
+    /// `data` is the path to the file.
+    /// `mode` is a hint how the file should be imported.
+    /// `progress` is a sender that provides a way for the importer to send progress messages
+    /// when importing large files. This also serves as a way to cancel the import. If the
+    /// consumer of the progress messages is dropped, subsequent attempts to send progress
+    /// will fail.
+    ///
+    /// Returns the hash of the imported file. The reason to have this method is that some database
+    /// implementations might be able to import a file without copying it.
+    fn import(
+        &self,
+        data: PathBuf,
+        mode: ImportMode,
+        progress: impl ProgressSender<Msg = ImportProgress> + IdGenerator,
+    ) -> BoxFuture<'_, io::Result<(Hash, u64)>>;
+
+    /// This trait method imports data from memory.
+    ///
+    /// It is a special case of `import` that does not use the file system.
+    fn import_bytes(&self, bytes: Bytes) -> BoxFuture<'_, io::Result<Hash>>;
+}
+
+/// Progress messages for an import operation
+///
+/// An import operation involves computing the outboard of a file, and then
+/// either copying or moving the file into the database.
+#[allow(missing_docs)]
+#[derive(Debug)]
+pub enum ImportProgress {
+    /// Found a path
+    ///
+    /// This will be the first message for an id
+    Found { id: u64, path: PathBuf },
```

Comment:
what does an `id` reference? a blob? or is this an event id?

Reply:
This is just an arbitrary id that allows you to save space in subsequent and very frequent progress calls like the update... ones. You wouldn't want to serialize the entire path in every update call, otherwise sending the update events over the wire would take more time than the actual process you are reporting progress on.